### ISSUE TITLE
[QUESTION] - How to get Listener from server.Default() so that I could use it in different libraries?

### ISSUE THREAD

**Issue by sujit-baniya** [cloudwego/hertz#545]
**Describe the Question**

I'm trying to use `hertz` framework with `https://github.com/pedia/endless` to allow zero downtime restart. There's an example to include Listener but I'm not sure how to use it with `hertz`

**Reproducible Code**

```go
func serve_http(addr string, ln net.Listener) (*fasthttp.Server, error) {
	server := &fasthttp.Server{Handler: func(ctx *fasthttp.RequestCtx) {
		ctx.WriteString(fmt.Sprintf("%d\n", os.Getpid()))
	}}

	go server.Serve(ln)

	return server, nil
}

func main() {
	addr := ":3030"
	var s *fasthttp.Server

	endless.Start(
		func(p *endless.Parent) error {
			ln, err := net.Listen("tcp", addr)
			if err != nil {
				return err
			}

			p.AddListener(ln, addr)
			s, err = serve_http(addr, ln)
			return err
		}, func(c *endless.Child) error {
			nf, ok := c.NamedFiles[addr]
			fmt.Printf("got %#v\n", nf)
			if !ok {
				return fmt.Errorf("inherit %s not found", addr)
			}

			c.AddListener(nf.LN, addr)
			server, err := serve_http(addr, nf.LN)
			s = server
			return err
		},
		func(ctx context.Context) error {
			return s.Shutdown()
		},
	)
	fmt.Printf("Quit %d\n", os.Getpid())
}

```

**Expected behavior**

Able to get listener

**Screenshots**

If applicable, add screenshots to help explain your question.

**Hertz version:**

Latest

**Comment by li-jin-gou** [comment]
I will research it and welcome it together.💕

**Comment by sujit-baniya** [comment]
@li-jin-gou Thanks. Please suggest if you've any idea

**Comment by welkeyever** [comment]
Seems that you wanna pass a customized listener to Hertz?
Actually it doesn't support that now, but I will looking into it - there isn't much work to do.

**Comment by sujit-baniya** [comment]
@welkeyever No, I want to get a listener from Hertz and use it for different purposes. So basically when Hertz is set up for `Spin` if it could store Listener somewhere and I could access that Listener from `RequestContext`, then that would also work.

**Comment by welkeyever** [comment]
@sujit-baniya It's reasonable to store it in network layer of hertz instance - you may get it through transport-related APIs. But it would be strange that relate listener to `RequestContext` - `RequestContext` is main for application-related jobs - you may save hertz instance somewhere, and get it to do whatever you wanna later.

**Comment by sujit-baniya** [comment]
@welkeyever Yes getting listener from hertz instance would also work. I was giving `RequestContext` as an example :)

**Comment by sujit-baniya** [comment]
Anything like `hertzInstance.GetListener()` or `hertzInstance.Listener` would work.

**Comment by welkeyever** [comment]
Okay, let me have a try. 
I will update if there is any progress.

**Comment by sujit-baniya** [comment]
@welkeyever any thought on this issue?

**Comment by welkeyever** [comment]
> @welkeyever any thought on this issue?

@sujit-baniya Sorry for the late. I was stumped by other things before. I will continue with this part.

**Comment by welkeyever** [comment]
Possible solution goes here: https://github.com/cloudwego/hertz/compare/develop...welkeyever:hertz:feat/more_options_for_transporter

Please feel free to add your review comments.

The endless example works, however there is a issue related to a third party lib:
![image](https://user-images.githubusercontent.com/9826697/229106865-f8d5a069-c6e7-4f4f-9c0b-acb7803a9871.png)

And we will finally get rid of this lib after this #541.

**Comment by sujit-baniya** [comment]
@welkeyever Please let me know your thought if we use Options.Listener net.Listener for Engine and pass it to transporter.

Something like this:

```go
type Options struct {
    ....
    Listener net.Listener
}

func WithListener(ln net.Listener) config.Option {
	return config.Option{F: func(o *config.Options) {
		o.Listener = ln
	}}
}
```
And for each transport we assign this listener to transport

```go
// For transporter switch
func NewTransporter(options *config.Options) network.Transporter {
	return &transport{
		readBufferSize:   options.ReadBufferSize,
		network:          options.Network,
		addr:             options.Addr,
		keepAliveTimeout: options.KeepAliveTimeout,
		readTimeout:      options.ReadTimeout,
		tls:              options.TLS,
		listenConfig:     options.ListenConfig,
                ln: options.Listener,
		OnAccept:         options.OnAccept,
		OnConnect:        options.OnConnect,
	}
}
```
And in `transport.serve()`

```go
    if t.ln == nil {
    network.UnlinkUdsFile(t.network, t.addr) //nolint:errcheck
	t.lock.Lock()
	if t.listenConfig != nil {
		t.ln, err = t.listenConfig.Listen(context.Background(), t.network, t.addr)
	} else {
		t.ln, err = net.Listen(t.network, t.addr)
	}
	t.lock.Unlock()
	if err != nil {
		return err
	}
    }
```

**Comment by sujit-baniya** [comment]
I did above and got the same error as yours

**Comment by welkeyever** [comment]
> ```go
> WithListener
> ```

Continuously adding options like `WithListener` will make it confusing when user adding both `WithListenConfig` and `WithListener` - there is a potential effective priority under this scenario, but it cannot be well informed to the user.

So I just provider a config-independent way to do so.

**Comment by 0x5487** [comment]
@welkeyever 
I also need to get listener/set listener from hertz.  I was wondering if the feature will be available in Hertz?  or we need to folk?  Thank you

**Comment by bjdgyc** [comment]
```
if t.listenConfig != nil {
    t.listener, err = t.listenConfig.Listen(context.Background(), t.network, t.addr)
} else {
    t.listener, err = net.Listen(t.network, t.addr)
}
```
以上为 hertz 的代码，`listenConfig` 直接调用了 `Listen`  方法，并没有进行特殊处理。
我认为框架内只需传入 `Listener net.Listener` 参数，不应传入 `listenConfig *net.ListenConfig` 参数，让用户去自定义 `net.Listener`。

代码参考：
```
t.lock.Lock()
if t.ln == nil {
   t.listener, err = net.Listen(t.network, t.addr)
}
t.lock.Unlock()
```
